[WIP] Add preliminary Clickhouse formatter

mattbasta opened the pull request:
Addresses https://github.com/sql-formatter-org/sql-formatter/issues/614

I've started putting together a Clickhouse formatter. Before I start getting into the weeds with tests and updating all of the docs/tests/playground, I wanted to check that things are directionally correct. Please let me know if there's anything you'd like to see done differently before I proceed.

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,161 @@
+      // https://clickhouse.com/docs/sql-reference/syntax#defining-and-using-query-parameters
+      custom: [
+        {
+          regex: String.raw`\{\s*[a-zA-Z0-9_]+\s*:\s*[a-zA-Z0-9_]+\s*\}`,
```
That regex would fail when a more complex type is used in the parameter, like this:

```sql
SELECT {d: Map(String, Array(UInt8))};
```

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,161 @@
+  'UPDATE',
+  // https://clickhouse.com/docs/sql-reference/statements/system
+  'SYSTEM RELOAD {DICTIONARIES | DICTIONARY | FUNCTIONS | FUNCTION | ASYNCHRONOUS METRICS} [ON CLUSTER]',
+  'SYSTEM DROP {DNS CACHE | MARK CACHE | ICEBERG METADATA CACHE | TEXT INDEX DICTIONARY CACHE | TEXT INDEX HEADER CACHE | TEXT INDEX POSTINGS CACHE | REPLICA | DATABASE REPLICA | UNCOMPRESSED CACHE | COMPILED EXPRESSION CACHE | QUERY CONDITION CACHE | QUERY CACHE | FORMAT SCHEMA CACHE | DROP FILESYSTEM CACHE}',
```
`SYSTEM DROP {... | DROP FILESYSTEM CACHE}`

will result in:

`SYSTEM DROP DROP FILESYSTEM CACHE`

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,161 @@
+  'DROP {DICTIONARY | DATABASE | USER | ROLE | QUOTA | PROFILE | SETTINGS PROFILE | VIEW | FUNCTION | NAMED COLLECTION | ROW POLICY | POLICY} [IF EXISTS]',
+  'DROP [TEMPORARY] TABLE [IF EXISTS] [IF EMPTY]',
+  // https://clickhouse.com/docs/sql-reference/statements/exists
+  'EXISTS [TEMPORARY] {TABLE | DICTIONARY | DATABASE}',
```
The linked docs say that the `TABLE | DICTIONARY | DATABASE` part is optional.

But making all that optional would result in `EXISTS (...)` expressions being detected as `EXISTS` statements.

So I guess you made a deliberate choice here.

nene commented on src/languages/clickhouse/clickhouse.keywords.ts:
```diff
@@ -0,0 +1,782 @@
+
+export const keywordPhrases: string[] = [
+  // See documentation of `keywords` above
+  'ADD COLUMN',
```
These should really be just part of keywords.

The -phrases variables should be used for the cases where you're using the `expandPhrases()` function to expand expressions like `TIMESTAMP {WITH | WITHOUT} TIME ZONE`.

When you just need multi-word keywords, they work just fine out of the box. Just like you're using multi-word data type names further down.

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,161 @@
+const reservedClauses = expandPhrases([
+  // https://clickhouse.com/docs/sql-reference/statements/explain
+  'EXPLAIN [AST | SYNTAX | QUERY TREE | PLAN | PIPELINE | ESTIMATE | TABLE OVERRIDE]',
+]);
```
Also, I don't expect to see the `EXPLAIN` statement in here - for all other dialects the `EXPLAIN` statement has been listed under `tabularOnelineClauses`.

mattbasta commented:
@nene I appreciate you taking the time to look! I'm very aware that it's not nearly ready to land and that it needs quite a bit of love, I just didn't want to invest a ton more time if it's completely off the mark. I have a commit in progress with the tests; I'm trying to aim for nearly all of the example queries from the docs to get handled correctly. I'll ping you when I have something that's ready for review.

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,362 @@
+ * when they are used as operators (not function calls).
+ *
+ * IN operator: foo IN (1, 2, 3) - IN comes after an identifier/expression
+ * IN function: IN(foo, 1, 2, 3) - IN comes at start or after operators/keywords
```
I see `in` listed in section titled [Functions for Implementing the IN Operator](https://clickhouse.com/docs/sql-reference/functions/in-functions)

This gives me the impression that `in` being available as a function is more of an implementation detail of Clickhouse and one normally wouldn't use it like so.

nene commented on src/languages/clickhouse/clickhouse.formatter.ts:
```diff
@@ -0,0 +1,362 @@
+    // Check text (uppercase canonical form) for matching, but preserve raw (original casing)
+    if (
+      token.type === TokenType.RESERVED_FUNCTION_NAME &&
+      (token.text === 'IN' || token.text === 'ANY')
```
I see `IN` and `ANY` being listed in both `clickhouse.keywords.ts` and `clickhouse.functions.ts`. There seems to be many more. Like there's `AND`, `OR`, `NOT`, `CAST`, `DATE` in both functions and keywords.

Having them listed in both places creates confusion, as it's not obvious which one takes priority. It happens to be that function names are detected before keywords by the tokenizer, so these all end up being classified as function names. But it would be better to avoid this ambiguity by only listing them in one place.

I would suggest listing them in the category that's the more common use case. Like I would move all these things to the keywords list and remove them completely from functions. This way it also aligns better with other dialects.

PS.  `AND` and `OR` won't really have any effect in either list as these are hard-coded to the tokenizer and will always get detected as special type of tokens.

mattbasta commented:
Hey @nene, apologies for the delay in following up. It's been a busy month here.

I think I've addressed all of your comments. Please let me know if there's anything that you'd like changed or that you aren't happy with in the current implementation.

nene commented:
Thanks for all the work. Sorry for the delay in responding, but it looks all good now.

Also thanks for updating the Github workflow file. (FYI, I didn't at all expect you to do this, but it's great that you did.)
